File: 3-Structural Patterns/1-Bridge Pattern/Customer Billing Example/Python/customer.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Customer module.
"""

__author__ = 'Ziang Lu'

from abc import ABC

from billing_implementor import BillingImplementor


class Customer(ABC):
    """
    Abstract Shape class that works as "Abstraction".
    This class defines an add_drinks() method for the client to call, and holds
    a reference to a billing implementor, so that the client will use that
    implementor to calculate the actual price for drinks.
    """
    __slots__ = ['_curr_billing_implementor', '_rounds']

    def __init__(self, billing_implementor: BillingImplementor):
        """
        Constructor with parameter.
        :param billing_implementor: BillingImplementor
        """
        self._curr_billing_implementor = billing_implementor
        self._rounds = []

    def set_billing_implementor(
        self, billing_implementor: BillingImplementor
    ) -> None:
        """
        Mutator of billing_implementor.
        :param billing_implementor: BillingImplementor
        :return: None
        """
        self._curr_billing_implementor = billing_implementor

    def add_drinks(self, price: float, n: int) -> None:
        """
        Adds the given number of the given drinks.
        :param price: float
        :param n: int
        :return: None
        """
        self._rounds.append(
            self._curr_billing_implementor.get_actual_price(price * n)
        )

    def print_bill(self) -> None:
        """
        Prints the total bill for this customer.
        :return: None
        """
        total = sum(self._rounds)
        print(f'Total due: {total}')
        self._rounds.clear()


class NormalCustomer(Customer):
    """
    Concrete NormalCustomer class that works as "ConcreteImplementor".
    """
    __slots__ = []

    def __init__(self, billing_implementor: BillingImplementor):
        """
        Constructor with parameter.
        :param billing_implementor: BillingImplementor
        """
        super().__init__(billing_implementor)
        self._rounds = []


class TeenagerCustomer(Customer):
    """
    Concrete TeenagerCustomer class that works as "ConcreteImplementor".
    """
    __slots__ = []

    def __init__(self, billing_implementor: BillingImplementor):
        """
        Constructor with parameter.
        :param billing_implementor: BillingImplementor
        """
        super().__init__(billing_implementor)
        self._rounds = []

    def add_drinks(self, price, n):
        if len(self._rounds) < 3:
            super().add_drinks(price, n)

File: 3-Structural Patterns/1-Bridge Pattern/Customer Billing Example/Java/customer/Customer.java
package customer;

import billing_implementor.BillingImplementor;

import java.util.ArrayList;
import java.util.List;

/**
 * Abstract Shape class that works as "Abstraction".
 * This class defines an addDrinks() method for the client to call, and holds a
 * reference to a billing implementor, so that the client will use that
 * implementor to calculate the actual price for drinks.
 *
 * @author Ziang Lu
 */
public abstract class Customer {

    /**
     * Current billing implementor.
     */
    private BillingImplementor currBillingImplementor;
    /**
     * List of drink rounds.
     */
    protected List<Double> rounds;

    /**
     * Constructor with parameter.
     * @param billingImplementor billing implementor to use
     */
    protected Customer(BillingImplementor billingImplementor) {
        currBillingImplementor = billingImplementor;
        rounds = new ArrayList<>();
    }

    /**
     * Mutator of billing implementor.
     * @param billingImplementor billing implementor to set
     */
    public void setBillingImplementor(BillingImplementor billingImplementor) {
        currBillingImplementor = billingImplementor;
    }

    /**
     * Adds the given number of the given drinks.
     * @param price price per drink
     * @param n number of rounds
     */
    public void addDrinks(double price, int n) {
        rounds.add(currBillingImplementor.getActualPrice(price * n));
    }

    /**
     * Prints the total bill for this customer.
     */
    public void printBill() {
        double sum = rounds.stream().mapToDouble(Double::doubleValue).sum();
        System.out.println("Total due: " + sum);
        rounds.clear();
    }

}
